src/java/org/apache/cassandra/db/virtual/SettingsTable.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.ColumnIdentifier;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.marshal.BooleanType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.partitions.PartitionUpdate;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.db.virtual.AbstractVirtualTable.DataSet;
+import org.apache.cassandra.dht.LocalPartitioner;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.StorageProxy;
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.transport.ServerError;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+
+final class SettingsTable extends AbstractVirtualTable
+{
+    private static final String VALUE = "value";
+    private static final String SETTING = "setting";
+
+    private static final List<Field> FIELDS = Arrays.stream(Config.class.getFields())
+            .filter(f -> !Modifier.isStatic(f.getModifiers()))
+            .collect(Collectors.toList());
+
+    SettingsTable(String keyspace)
+    {
+        super(TableMetadata.builder(keyspace, "settings")
+                .comment("current settings")
+                .kind(TableMetadata.Kind.VIRTUAL)
+                .partitioner(new LocalPartitioner(UTF8Type.instance))
+                .addPartitionKeyColumn(SETTING, UTF8Type.instance)
+                .addRegularColumn(VALUE, UTF8Type.instance)
+                .build());
+    }
+
+    public DataSet data(DecoratedKey partitionKey)
+    {
+        SimpleDataSet result = new SimpleDataSet(metadata());
+        Config config = DatabaseDescriptor.getRawConfig();
+        String setting = null;
+        try
+        {
+            setting = UTF8Type.instance.compose(partitionKey.getKey());
+            Field field = config.getClass().getDeclaredField(setting);
+            if (!FIELDS.contains(field))
+                throw new InvalidRequestException("Invalid setting: " + setting);
+            Object value = field.get(config);
+            result.row(setting);
+            if (value != null)
+                result.column(VALUE, value.toString());
+        }
+        catch (IllegalAccessException | IllegalArgumentException e)
+        {
+            throw new ServerError(e);
+        }
+        catch (NoSuchFieldException e)
+        {
+            throw new InvalidRequestException("No such setting: " + setting);
+        }
+        return result;
+    }
+
+    @Override
+    public DataSet data()
+    {
+        SimpleDataSet result = new SimpleDataSet(metadata());
+        Config config = DatabaseDescriptor.getRawConfig();
+        for (Field f : FIELDS)
+        {
+            try
+            {
+                Object value = f.get(config);
```

Comment:
Should probably factor out this whole block into a method usable by both `data()` implementations. ` String fieldValue(Field field, Config config)`?